peps/pep-0749.rst

```diff
@@ -0,0 +1,741 @@
+
+Third-party code that implements ``__annotate__`` functions should either
+support all three annotation formats, or be prepared to handle the
+"fake globals" environment. This should be mentioned in the data model
+documentation for ``__annotate__``.
+
+Effect of setting ``__annotations__``
+=====================================
+
+:pep:`649` specifies:
+
+  Setting ``o.__annotations__`` to a legal value
+  automatically sets ``o.__annotate__`` to ``None``.
+
+We would prefer to keep ``__annotate__`` unchanged when ``__annotations__``
+is written to. Conceptually, ``__annotate__`` provides the ground truth
+and ``__annotations__`` is merely a cache, and we shouldn't throw away the
+ground truth if the cache is modified.
+
+The motivation for :pep:`649`'s behavior is to keep the two attributes in sync.
+However, this is impossible in general; if the ``__annotations__`` dictionary
+is modified in place, this will not be reflected in the ``__annotate__`` attribute.
+The overall mental model for this area will be simpler if setting ``__annotations__``
+has no effect on ``__annotate__``.
+
+Specification
+-------------
+
+The value of ``__annotate__`` is not changed when ``__annotations__`` is set.
+
+Deferred evaluation of PEP 695 and 696 objects
+==============================================
+
+Since :pep:`649` was written, Python 3.12 and 3.13 gained support for
+several new features that also use deferred evaluation, similar to the
+behavior this PEP proposes for annotations:
+
+* The value of type aliases created through the :py:keyword:`type`
+  statement (:pep:`695`)
+* The bound and constraints of :py:class:`typing.TypeVar` objects
+  created through the syntax for generics (:pep:`695`)
+* The default value of :py:class:`typing.TypeVar`, :py:class:`ParamSpec`,
+  and :py:class:`typing.TypeVarTuple` objects (:pep:`696`)
+
+Currently, these objects use deferred evaluation, but there is no direct
+access to the function object used for deferred evaluation. To enable
+the same kind of introspection that is now possible for annotations, we propose
+to expose the internal function objects, allowing users to evaluate them
+using the FORWARDREF and SOURCE formats.
+
+Specification
+-------------
+
+We will add the following new attributes:
+
+* ``evaluate_value`` on :py:class:`typing.TypeAliasType`
+* ``evaluate_bound``, ``evaluate_constraints``, and ``evaluate_default`` on :py:class:`typing.TypeVar`
+* ``evaluate_default`` on :py:class:`typing.ParamSpec`
+* ``evaluate_default`` on :py:class:`typing.TypeVarTuple`
+
+Except for ``evaluate_value``, these attributes may be ``None`` if the object
+does not have a bound, constraints, or default. Otherwise, the attribute is a
+callable, similar to an ``__annotate__`` function, that takes a single integer
+argument and returns the evaluated value. Unlike ``__annotate__`` functions,
+these callables return a single value, not a dictionary of annotations.
+These attributes are read-only.
```

Comment:
Would it make more sense for `evaluate_default` to always be set to a callable, and have that callable simply return `NoDefault` if there is no default?

Similarly, `evaluate_bound` could just return `None` if there is no bound, and `evaluate_constraints` could just return an empty tuple if there are no constraints

Reply:
Possibly, it feels cleaner to me to have the functions not exist at all if there's nothing to evaluate, but could definitely go either way.

One advantage of the current approach is that it allows you to distinguish between a TypeVar with no bound and one with a TypeVar set to None, which is more of a theoretical concern than a real one, but it feels nice that we can at least represent the distinction.

Reply:
> Hrm, to me it feels like it would be a pain if I'd have to explicitly check that the attribute is callable before I called it

You'd usually call it through `annotations.call_evaluate_function`, and that handles None explicitly.

Reply:
Why are we exposing them publicly if we don't want users calling them directly? Would it be better to expose them using private names?

Reply:
We expect people to be using them, just not to call them. If we make them private, would we add a separate functions to `annotations` to get a TypeVar bound, type alias value, etc.?

Reply:
Oh I see what you're saying. And it is clearly specified that this is the intended use case. Sorry for being slow!

It does still feel a _bit_ strange to say that we'll be exposing a new public method on this object, but we'd prefer it if you didn't call it directly.

Reply:
It's not too different from the `__annotate__` attribute we're exposing on functions, modules, and classes: you usually shouldn't call that directly either.

Reply:
Right, but that one's a dunder, so the name already sort-of signals that it's at least doing something pretty special